src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs

```diff
@@ -1332,80 +1361,58 @@ void CleanExceptionHandlers ()
-			{
-				switch (instr.OpCode.Code) {
-				case Code.Ldarg:
-				case Code.Ldloc:
-				case Code.Ldloc_0:
-				case Code.Ldloc_1:
-				case Code.Ldloc_2:
-				case Code.Ldloc_3:
-				case Code.Ldloc_S:
-				case Code.Ldc_I4_0:
-				case Code.Ldc_I4_1:
-				case Code.Ldc_I4_2:
-				case Code.Ldc_I4_3:
-				case Code.Ldc_I4_4:
-				case Code.Ldc_I4_5:
-				case Code.Ldc_I4_6:
-				case Code.Ldc_I4_7:
-				case Code.Ldc_I4_8:
-				case Code.Ldc_I4:
-				case Code.Ldc_I4_S:
-				case Code.Ldc_I4_M1:
-				case Code.Ldc_I8:
-				case Code.Ldc_R4:
-				case Code.Ldc_R8:
-				case Code.Ldnull:
-				case Code.Ldstr:
-					return true;
-				}
-
-				return false;
-			}
 		}
 
 		struct ConstantExpressionMethodAnalyzer
 		{
-			readonly MethodDefinition method;
-			readonly Collection<Instruction> instructions;
 			readonly LinkContext context;
+			readonly UnreachableBlocksOptimizer optimizer;
 
 			Stack<Instruction>? stack_instr;
 			Dictionary<int, Instruction>? locals;
 
-			public ConstantExpressionMethodAnalyzer (LinkContext context, MethodDefinition method)
+			public ConstantExpressionMethodAnalyzer (UnreachableBlocksOptimizer optimizer)
 			{
-				this.context = context;
-				this.method = method;
-				instructions = method.Body.Instructions;
+				this.optimizer = optimizer;
+				this.context = optimizer._context;
 				stack_instr = null;
 				locals = null;
 				Result = null;
+				SideEffectFreeResult = true;
 			}
 
-			public ConstantExpressionMethodAnalyzer (LinkContext context, MethodDefinition method, Collection<Instruction> instructions)
-				: this (context, method)
-			{
-				this.instructions = instructions;
-			}
-
+			//
+			// Single expression that is representing the evaluation result with the specific
+			// callee arguments
+			//
 			public Instruction? Result { get; private set; }
 
+			//
+			// Returns true when the method evaluation with specific arguments does not cause
+			// any observable side effect (e.g. possible NRE, field access, etc)
+			//
+			public bool SideEffectFreeResult { get; private set; }
+
 			[MemberNotNullWhen (true, "Result")]
-			public bool Analyze ()
+			public bool Analyze (in CalleePayload callee, Stack<MethodDefinition> callStack)
 			{
-				var body = method.Body;
-				if (body.HasExceptionHandlers)
-					return false;
+				MethodDefinition method = callee.Method;
+				Instruction[]? arguments = callee.Arguments;
+				Collection<Instruction> instructions = callee.Method.Body.Instructions;
+				MethodBody body = method.Body;
 
 				VariableReference vr;
 				Instruction? jmpTarget = null;
 				Instruction? linstr;
+				object? left, right, operand;
+
+				//
+				// We could implement a full-blown interpreter here but for now, it handles
+				// cases used in runtime libraries
+				//
+				for (int i = 0; i < instructions.Count; ++i) {
+					var instr = instructions[i];
 
-				foreach (var instr in instructions) {
 					if (jmpTarget != null) {
 						if (instr != jmpTarget)
 							continue;
```

Comment:
This is a weird way to handle backward branches. Maybe add a comment that this will effectively `break` on a backward jump.

Reply:
commend added